dbms/src/Storages/DeltaMerge/VersionChain/DMFileHandleIndex.h

```diff
@@ -0,0 +1,241 @@
+            max_values.begin() + clipped_pack_range.start_pack_id,
+            max_values.begin() + clipped_pack_range.end_pack_id);
+    }
+
+    std::vector<UInt32> loadPackOffsets() const
+    {
+        const auto & pack_stats = dmfile->getPackStats();
+        std::vector<UInt32> pack_offsets(clipped_pack_range.count(), 0);
+        for (UInt32 clipped_pack_id = 1; clipped_pack_id < pack_offsets.size(); ++clipped_pack_id)
+        {
+            UInt32 real_pack_id = clipped_pack_id + clipped_pack_range.start_pack_id;
+            pack_offsets[clipped_pack_id] = pack_offsets[clipped_pack_id - 1] + pack_stats[real_pack_id - 1].rows;
+        }
+        return pack_offsets;
+    }
+
+    void loadHandleIfNotLoaded(const DMContext & dm_context)
+    {
+        if (likely(clipped_need_read_packs.empty()))
+            return;
+
+        auto read_pack_ids = std::make_shared<IdSet>();
+        for (UInt32 i = 0; i < clipped_need_read_packs.size(); ++i)
+        {
+            if (clipped_need_read_packs[i])
+                read_pack_ids->insert(i + clipped_pack_range.start_pack_id);
+        }
+
+        auto pack_filter = DMFilePackFilter::loadFrom(
+            dm_context,
+            dmfile,
+            /*set_cache_if_miss*/ true,
+            /*rowkey_ranges*/ {}, // Empty means all
+            EMPTY_RS_OPERATOR,
+            read_pack_ids);
+
+        DMFileReader reader(
+            dmfile,
+            {getHandleColumnDefine<HandleType>()},
+            isCommonHandle<HandleType>(),
+            /*enable_handle_clean_read*/ false,
+            /*enable_del_clean_read*/ false,
+            /*is_fast_scan*/ false,
+            /*max_data_version*/ std::numeric_limits<UInt64>::max(),
+            std::move(pack_filter),
+            dm_context.global_context.getMarkCache(),
+            /*enable_column_cache*/ false,
+            /*column_cache*/ nullptr,
+            dm_context.global_context.getSettingsRef().max_read_buffer_size,
+            dm_context.global_context.getFileProvider(),
+            dm_context.global_context.getReadLimiter(),
+            DEFAULT_MERGE_BLOCK_SIZE,
+            /*read_one_pack_every_time*/ true,
+            dm_context.tracing_id,
+            /*max_sharing_column_bytes_for_all*/ false,
+            dm_context.scan_context,
+            ReadTag::MVCC);
+
+        const auto & pack_stats = dmfile->getPackStats();
+        for (UInt32 pack_id : *read_pack_ids)
+        {
+            auto block = reader.read();
+            RUNTIME_CHECK(block.rows() == pack_stats[pack_id].rows, pack_id, block.rows(), pack_stats[pack_id].rows);
+            clipped_handle_packs[pack_id - clipped_pack_range.start_pack_id] = block.begin()->column;
+        }
+        clipped_need_read_packs.clear();
+    }
+
+    struct PackRange
+    {
+        // [start_pack_id, end_pack_id)
+        UInt32 start_pack_id;
+        UInt32 end_pack_id;
+
+        UInt32 count() const { return end_pack_id - start_pack_id; }
+    };
```

Comment:
```suggestion
    struct PackRange
    {
        UInt32 start_pack_id;
        UInt32 count;
    };
```

`count` is more frequently used.